### PR TITLE
Add DexLabel support.

### PR DESCRIPTION
**Pull request by TomWeaver18**
Specify a DexLabel('myLabel') on any line of a test and you can now reference
said label within a DexExpectWatchValue('myVar', '1', '2', on_line='myLabel').

In order to preserve label values resolved at the point of parsing a command, the way in which we pass DexCommands around DExTer has needed to be re-jigged a little.

The original implementation would store a CommandIR with LocIR for each command, an abstraction of the raw text and the location information for each command. The raw text part was evaled throughout DExTer for different purposes, mainly in the run_internal_debugger tool and heuristic.

This is now no longer the case. DexCommands are now stored as objects and are passed around as such. The only exception being within the run_internal_debugger tool which loads them from previously pickled data.

As a result there are several changes in this patch that are there to support the use of command objects rather than CommandIR, which has now been removed from DExTer.

### CODE REVIEW

File: dexter.py [hunk 1]
```diff
@@ -27,7 +27,6 @@
 import sys
 
 from dex.tools import main
-from dex.utils.ReturnCode import ReturnCode
```
Comment (TomWeaver18) [inline dexter.py]: This is a stealth change, my linter was complaining about an unused import

File: dex/command/ParseCommand.py [hunk 2]
```diff
@@ -33,6 +33,19 @@
 from dex.command.CommandBase import CommandBase
 from dex.utils.Exceptions import CommandParseError
 
+def import_command_modules():
```
Comment (TomWeaver18) [inline dex/command/ParseCommand.py]: This function is called in the run_internal_debugger tool to import commands into the spawned child process.

There's some duplicate code in this function stolen from _get_valid_commands.

There'll be some clean up needed in the future to resolve these two functions.

I think we should follow the same logic as we did with the debugger and tool discovery code and just do this statically.

File: dex/heuristic/Heuristic.py [hunk 3]
```diff
@@ -158,8 +158,7 @@ def __init__(self, context, steps):
         try:
             penalties = defaultdict(list)
             maximum_possible_penalty_all = 0
-            for command in steps.commands["DexExpectProgramState"]:
-                expect_state = get_command_object(command)
+            for expect_state in steps.commands["DexExpectProgramState"]:
```
Comment (OCHyams) [inline dex/heuristic/Heuristic.py]: Nit: Should be ' instead of ".

File: dex/command/CommandBase.py [hunk 4]
```diff
@@ -32,6 +32,12 @@ def __init__(self):
         self.path = None
         self.lineno = None
 
+    def get_label_args(self):
+        return list()
+
+    def has_labels(self):
```
[the remark this answers is not in the document]
Comment (OCHyams) [reply dex/command/CommandBase.py]: It's then one less thing to worry about (or forget, get wrong) when implementing new commands.

File: dex/command/commands/DexExpectWatchValue.py [hunk 5]
```diff
@@ -126,6 +126,32 @@ def missing_values(self):
     def encountered_values(self):
         return sorted(list(set(self.values) - self._missing_values))
 
+    def get_label_args(self):
+        label_list = []
+        if self.has_labels():
+            label_list = [self._to_line, self._from_line]
+        return label_list
+
+    def resolve_label(self, label_line_pair):
```
Comment (OCHyams) [inline dex/command/commands/DexExpectWatchValue.py]: The codebase is a little inconsistent on it, but I think we're trying to lean towards adding type annotations to functions. Sometimes the added friction of annotations isn't necessary but in those cases I reckon there should be a docstring explaining any interface requirements (e.g. param label_line_pair must be indexable).

File: dex/command/ParseCommand.py (hunk 2 above)
Comment (TomWeaver18) [reply dex/command/ParseCommand.py]: so I agree, but _get_valid_commands is called in other places in the code base and would need updating in those places too. In those places it makes sense to be called "_get_valid_commands" as their expecting a collection of command strings to validate parsed commands against.

if it's a must that the deduplication must go before going in, then I'd say that _get_valid_commands() needs reworking before this goes in as I see that as a seperate issue.

I didn't want a call to _get_valid_commands in the run_internal_debugger code as that would look rather hacky considering it's documented (by the prefixed '_') as a static function just for use in the ParseCommands.py file.

TL:DR, we should re-jig how imports of commands are done in Dexter to avoid having to do this kind of deduplication.

File: dex/command/CommandBase.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TomWeaver18) [reply dex/command/CommandBase.py]: so there's some code that calls get_label_args() in resolve_labels in ParseCommand.py, expecting a collection for use in a loop. will returning None at that point cause an issue?

File: dex/heuristic/Heuristic.py [hunk 6]
```diff
@@ -144,7 +144,7 @@ def __init__(self, context, steps):
         # Get DexExpectWatchValue results.
         try:
             for watch in steps.commands["DexExpectWatchValue"]:
-                command = get_command_object(watch)
+                command = watch
```
Comment (Melamoto) [inline dex/heuristic/Heuristic.py]: maybe just delete this line and rename watch -> command?

File: dex/command/CommandBase.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (OCHyams) [reply dex/command/CommandBase.py]: In which case you could return an empty list from get_label_args() if there are none (as it is at the moment) and then:
```suggestion
def has_labels(self):
    return len(self.get_label_args()) > 0
```

My problem is that these functions are highly coupled and I don't think they need to be.

File: dex/command/ParseCommand.py (hunk 2 above)
Comment (OCHyams) [reply dex/command/ParseCommand.py]: In which case I feel that, as you say;

> _get_valid_commands() needs reworking before this goes in

File: dex/command/CommandBase.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (OCHyams) [reply dex/command/CommandBase.py]: If you strongly disagree I can live with your implementation.

File: dex/heuristic/Heuristic.py (hunk 6 above)
Comment (TomWeaver18) [reply dex/heuristic/Heuristic.py]: yeh over looked this one, thanks for the spot.

File: dex/command/ParseCommand.py (hunk 2 above)
Comment (TomWeaver18) [reply dex/command/ParseCommand.py]: right choo are, we'll have to sort that one out next week then.

File: dex/command/commands/DexLabel.py [hunk 7]
```diff
@@ -20,14 +20,24 @@
 # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
-"""Serialization of a DExTer command embedded within one of the files under
-test.
+"""Command to instruct the debugger to inspect the value of some set of
```
Comment (TomWeaver18) [inline dex/command/commands/DexLabel.py]: This string is left over from a copy pasta of DexWatch

File: dex/command/ParseCommand.py [hunk 8]
```diff
@@ -65,7 +78,7 @@ def _get_valid_commands():
 
 
 def get_command_object(commandIR):
```
Comment (TomWeaver18) [inline dex/command/ParseCommand.py]: I think this function may be eligible for deletion now given that it expects commandIR

File: dex/command/CommandBase.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TomWeaver18) [reply dex/command/CommandBase.py]: I've reworked both the get_label_args and has_labels to try and break the coupling.

hope yeh like em!

File: dex/command/ParseCommand.py (hunk 8 above)
Comment (TomWeaver18) [reply dex/command/ParseCommand.py]: this has been removed on my "up-to-date" branch.

File: dex/command/commands/DexExpectWatchValue.py (hunk 5 above)
Comment (TomWeaver18) [reply dex/command/commands/DexExpectWatchValue.py]: added type notations.

File: dex/command/ParseCommand.py (hunk 2 above)
Comment (TomWeaver18) [reply dex/command/ParseCommand.py]: this is being address in a seperate PR

Comment (jmorse) [reply dex/command/ParseCommand.py]: [Ignoring this as I assume this hunk as the code vanishes shortly?]

File: dex/command/commands/DexExpectWatchValue.py [hunk 9]
```diff
@@ -126,6 +126,32 @@ def missing_values(self):
     def encountered_values(self):
         return sorted(list(set(self.values) - self._missing_values))
 
+    def get_label_args(self):
+        label_list = []
+        if self.has_labels():
+            label_list = [self._to_line, self._from_line]
```
Comment (jmorse) [inline dex/command/commands/DexExpectWatchValue.py]: This needs to filter for attributes that _are_ labels, someone might use a line number in one place, and a label somewhere else; both would be returned as labels in this implementation.

File: dex/debugger/DebuggerBase.py [hunk 10]
```diff
@@ -117,9 +117,9 @@ def _update_step_watches(self, step_info):
         try:
             # Iterate over all watches of the types named in watch_cmds
             for watch in towatch:
-                if (watch.loc.path == loc.path
-                        and watch.loc.lineno == loc.lineno):
-                    result = get_command_object(watch).eval(self)
+                if (watch.path == loc.path
+                        and watch.lineno == loc.lineno):
+                    result = watch.eval(self)
```
Comment (jmorse) [inline dex/debugger/DebuggerBase.py]: I'm assuming this is to do with shuffing command import stuff that's being addressed elsewhere?

File: dex/command/commands/DexExpectWatchValue.py (hunk 9 above)
Comment (TomWeaver18) [reply dex/command/commands/DexExpectWatchValue.py]: this has already been addressed in further work

File: dex/debugger/DebuggerBase.py (hunk 10 above)
Comment (TomWeaver18) [reply dex/debugger/DebuggerBase.py]: this is to do with the removal of the multiple evals. command objects are only evaled once, in ParseCommands.py

File: dex/command/commands/DexExpectWatchValue.py (hunk 9 above)
Comment (jmorse) [reply dex/command/commands/DexExpectWatchValue.py]: Ah, you've pushed that to your "add_DexLabel_command" branch; that's not going to be visible in this pull request unless you push that on top of your "add_DexLabel_command_pr" branch.

Comment (TomWeaver18) [reply dex/command/commands/DexExpectWatchValue.py]: I do understand, I was waiting for the command import stuff on master to go through before pushing new stuff to this branch.

**Comment by TomWeaver18** [comment]
looks as though something freak has happened with my branch here... the history seems to be totally messed up. Please hold off reviewing/commenting until I get this fixed.

thanks 
Tom W

**Comment by TomWeaver18** [comment]
History is correct now.

File: dex/command/commands/DexExpectWatchValue.py [hunk 11]
```diff
@@ -126,6 +126,32 @@ def missing_values(self):
...
+        if self._to_line == label_line_pair[0]:
+            self._to_line = label_line_pair[1]
+        if self._from_line == label_line_pair[0]:
+             self._from_line = label_line_pair[1]
+
+    def has_labels(self):
+        """Labels for _from_line and _to_line are any none to int convertable
+           string.
+        """
+        try:
```
[the remark this answers is not in the document]
Comment (jmorse) [reply dex/command/commands/DexExpectWatchValue.py]: Would be even more clear with instance now:

    return [x for x in (self._from_line, self._to_line) if not isinstance(x, int)]

**Comment by TomWeaver18** [comment]
As removal of DexWatch has gone in, this stuff is now broken... so am re-working this patch... again.

**Comment by TomWeaver18** [comment]
Patch is now merged with master, conflict resolved and bug* free.

File: dex/command/ParseCommand.py [hunk 12]
```diff
@@ -60,7 +61,7 @@ def _get_command_name(command_raw: str) -> str:
 
 def _merge_subcommands(command_name: str, valid_commands: dict) -> dict:
     """Return a dict which merges valid_commands and subcommands for
-    command_name.
+       command_name.
```
Comment (OCHyams) [inline dex/command/ParseCommand.py]: Very minor nit: In #39 I try to match the Google style docstrings which do not indent the lines after the first like this.
If you have any other changes to make before this patch lands please can you remove this indentation?

**Comment by TomWeaver18** [comment]
I agree, but I'd rather leave the implementation of such features for future work.  DexLabels are in, the infrastructure for using them is in, it just requires command implementations to specify how they use them.
